File: binarySearch-2/src/Problem1.java

//TC : O(logn) as binary search is used
//SC : O(1)

// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no
public class Problem1 {

    public int[] searchRange(int[] nums, int target) {
            int[] soln = new int[2];
        soln[0] = findFirstOccurence(nums, target);
        soln[1] = findLastOccurence(nums, target);
            return soln;
        }


    //method to find first occurence, the only change is if [mid] == target then we dont stop there.
    //We set end=mid-1 and discard the right half as we want to find the first occurence
        private int findFirstOccurence(int[] nums, int target){
            int index = -1;
            int start = 0;
            int end = nums.length - 1;
            while(start <= end)
            {
                int mid = start + (end-start)/2;

                if(nums[mid] >= target)
                    end = mid - 1;
                else
                    start = mid + 1;

                if(nums[mid] == target)
                    index = mid;
            }
            return index;
        }

        //method to find last occurence, the only change is if [mid] == target then we dont stop there.
    //We set start = mid+1 and discard the left half as we want to find the last occurence
        private int findLastOccurence(int[] nums, int target){
            int index = -1;
            int start = 0;
            int end = nums.length - 1;
            while(start <= end)
            {
                int mid = start + (end-start)/2;

                if(nums[mid] <= target)
                    start = mid + 1;
                else
                    end = mid - 1;

                if(nums[mid] == target)
                    index = mid;
            }
            return index;
        }
}

File: Problem_1.js
// Time Complexity : O(log(n)) - Binary search
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes 
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

// ## Problem 1: (https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/)

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var getLeftOccurence = (nums, target) => {
    let low = 0;
    let n = nums.length;
    let high = n - 1;
    while (low <= high) {
        let mid = Math.floor(low + ((high - low) / 2)); // To prevent integer overflow
        if (nums[mid] === target) {
            // If target is found and it is the first element of the array or the int on the left is smaller,
            // the leftmost occurence is found
            if (mid === 0 || nums[mid - 1] < target) {
                return mid;
            }
            high = mid - 1;
        }
        else if (target < nums[mid]) {
            // If target is less than the mid value, search left half
            high = mid - 1;
        } else {
            // Serach right half
            low = mid + 1;
        }
    }
    return -1;
}
var getRightOccurence = (nums, target, lowerBound) => {
    let low = lowerBound;
    let n = nums.length;
    let high = n - 1;
    while (low <= high) {
        let mid = Math.floor(low + ((high - low) / 2)); // To prevent integer overflow

        if (nums[mid] === target) {
            // If target is found and it is the last element of the array or the int on the right is greater,
            // the rightmost occurence is found
            if (mid === n - 1 || nums[mid + 1] > target) {
                return mid;
            }
            low = mid + 1;
        }
        // If target is less than the mid value, search left half
        else if (target < nums[mid]) {
            high = mid - 1;
        } else {
            // Serach right half
            low = mid + 1;
        }
    }
    return -1;
}
var searchRange = function (nums, target) {
    // If the array does not exists or does not contain any element, return [-1,-1]
    if (nums === null || nums.length === 0) {
        return [-1, -1];
    }

    let left = (getLeftOccurence(nums, target));
    // If left occurence is not found, the element is not present in the array, return [-1,-1]
    if (left === -1) {
        return [-1, -1];
    }
    // If left index is found, the right index will be between right and array end. 
    // So find target in that range only
    let right = (getRightOccurence(nums, target, left));
    return [left, right];
};